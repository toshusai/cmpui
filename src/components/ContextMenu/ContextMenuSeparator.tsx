import * as RadixContextMenu from "@radix-ui/react-context-menu";

export function ContextMenuSeparator() {
  return <RadixContextMenu.Separator className="context-menu-separator" />;
}
