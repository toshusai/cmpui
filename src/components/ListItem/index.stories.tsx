import type { Meta, StoryObj } from "@storybook/react";
import { List, ListItem } from ".";
import { FloatBox } from "../FloatBox";

const meta: Meta<typeof ListItem> = {
  title: "View/ListItem",
  component: ListItem,
};

export default meta;
type Story = StoryObj<typeof ListItem>;

export const Basic: Story = {
  render: function RenderBasic() {
    return (
      <List as="ol">
        <ListItem>Item1</ListItem>
        <ListItem>Item2</ListItem>
        <ListItem>Item3</ListItem>
      </List>
    );
  },
};

export const Selected: Story = {
  render: function RenderBasic() {
    return (
      <List as="ol">
        <ListItem selected>Item1</ListItem>
        <ListItem>Item2</ListItem>
        <ListItem>Item3</ListItem>
      </List>
    );
  },
};

export const Small: Story = {
  render: function RenderBasic() {
    return (
      <List as="ol">
        <ListItem size="S">Item1</ListItem>
        <ListItem size="S">Item2</ListItem>
        <ListItem size="S">Item3</ListItem>
      </List>
    );
  },
};

function CommandItem(props: {
  children?: React.ReactNode;
  shortcut?: string;
  selected?: boolean;
}) {
  return (
    <ListItem
      style={{
        width: 128,
      }}
      rounded
      selected={props.selected}
      size="S"
    >
      {props.children}
      {props.shortcut && (
        <span
          style={{
            marginLeft: "auto",
            fontSize: 10,
            color: props.selected ? "white" : "var(--cmpui-text2-color)",
          }}
        >
          {props.shortcut}
        </span>
      )}
    </ListItem>
  );
}

export const Commands: Story = {
  render: function RenderBasic() {
    return (
      <FloatBox>
        <List as="ol">
          <CommandItem shortcut="⌘ Z">Undo</CommandItem>
          <CommandItem shortcut="⇧ ⌘ Z">Redo</CommandItem>
          <CommandItem shortcut="⌘ X">Cut</CommandItem>
          <CommandItem selected shortcut="⌘ C">
            Copy
          </CommandItem>
          <CommandItem shortcut="⌘ V">Paste</CommandItem>
        </List>
      </FloatBox>
    );
  },
};
