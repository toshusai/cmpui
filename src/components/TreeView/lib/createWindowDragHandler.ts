export function createWindowDragHandler<T, P>({
  onDown,
  onMove,
  onUp,
}: {
  onDown?: (e: React.PointerEvent<HTMLElement>) => T;
  onMove?: (e: PointerEvent, ctx?: T, prev?: P) => P;
  onUp?: (e: PointerEvent, ctx?: T, prev?: P) => void;
}) {
  const handlePointerDown = (downEvent: React.PointerEvent<HTMLElement>) => {
    if (downEvent.buttons === 2) {
      return;
    }
    const result = onDown?.(downEvent);
    let prev: P | undefined;
    const handlePointerMove = (e: PointerEvent) => {
      prev = onMove?.(e, result, prev);
    };

    const handlePointerUp = (e: PointerEvent) => {
      window.removeEventListener("pointermove", handlePointerMove);
      window.removeEventListener("pointerup", handlePointerUp);
      onUp?.(e, result, prev);
    };

    window.addEventListener("pointermove", handlePointerMove);
    window.addEventListener("pointerup", handlePointerUp);
  };

  return handlePointerDown;
}
