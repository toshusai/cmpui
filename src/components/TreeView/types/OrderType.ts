export enum OrderType {
  Before = 0,
  After = 1,
  Child = 2,
}
