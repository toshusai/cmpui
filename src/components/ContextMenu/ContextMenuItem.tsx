import * as RadixContextMenu from "@radix-ui/react-context-menu";
import { ListItem } from "../ListItem";

export type ContextMenuItemProps = React.ComponentProps<
  typeof RadixContextMenu.Item
>;

export function ContextMenuItem({ children, ...props }: ContextMenuItemProps) {
  return (
    <RadixContextMenu.Item {...props} asChild>
      <ListItem size="S" rounded>
        {children}
      </ListItem>
    </RadixContextMenu.Item>
  );
}
