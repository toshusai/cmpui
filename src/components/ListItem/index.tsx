import React, { ForwardedRef, forwardRef } from "react";
import { CustomJSXElement } from "../../types/CustomJSXElement";
import { classNames } from "../../utils/classNames";

import "./index.css";

export type ListItemProps<T extends CustomJSXElement = "li"> = {
  as?: T;
  children: React.ReactNode;
  className?: string;
  rounded?: boolean;
  selected?: boolean;
  size?: "S" | "M";
} & React.ComponentProps<T>;

export const ListItem = forwardRef(function ListItem<
  T extends CustomJSXElement,
>(
  {
    as: Component = "li",
    children,
    className,
    rounded,
    size = "M",
    ...props
  }: ListItemProps<T>,
  ref: ForwardedRef<HTMLElement>,
) {
  return (
    <Component
      className={classNames("cmpui_list-item__root", className)}
      ref={ref}
      aria-selected={props.selected}
      data-size={size}
      data-rounded={rounded}
      {...props}
    >
      {children}
    </Component>
  );
}) as <T extends CustomJSXElement = "li">(p: ListItemProps<T>) => JSX.Element;

export const List = forwardRef<
  HTMLUListElement,
  React.ComponentProps<"ul"> & {
    as?: React.ElementType;
  }
>(({ className, children, as: Component = "ul", ...props }, ref) => {
  return (
    <Component
      className={classNames("cmpui_list-list__root", className)}
      ref={ref}
      {...props}
    >
      {children}
    </Component>
  );
});
