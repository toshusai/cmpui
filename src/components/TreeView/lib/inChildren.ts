import { ITree } from "../../../types";

export function inChildren(item: ITree, id: string): boolean {
  if (item.id === id) return true;
  if (item.children) {
    for (const child of item.children) {
      if (inChildren(child, id)) return true;
    }
  }
  return false;
}
