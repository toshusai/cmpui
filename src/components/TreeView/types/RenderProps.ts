import { ITree } from "../../../types";

export type RenderProps<T extends ITree> = {
  depth: number;
  item: T;
  onOpenChange: (isOpen: boolean) => void;
  open: boolean;
};
