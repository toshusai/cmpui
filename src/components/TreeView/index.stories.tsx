import type { Meta, StoryObj } from "@storybook/react";
import { useState } from "react";
import { TreeView } from ".";
import { Icon } from "../../__stories__/Icon";
import { ITree } from "../../types";
import { ContextMenu } from "../ContextMenu";
import { ContextMenuItem } from "../ContextMenu/ContextMenuItem";
import { ContextMenuSeparator } from "../ContextMenu/ContextMenuSeparator";
import { ListItem } from "../ListItem";
import { inChildren } from "./lib/inChildren";
import { OrderType } from "./types/OrderType";
import { RenderProps } from "./types/RenderProps";

const meta: Meta<typeof TreeView> = {
  title: "View/TreeView",
  component: TreeView,
};

export default meta;

type Story = StoryObj<typeof TreeView>;

function randomId() {
  return Math.random().toString(36).slice(2);
}

type Item = {
  id: string;
  children?: Item[];
  icon?: string;
};

const defaultItems: Item[] = [
  {
    id: "root",
    children: [
      {
        id: "src",
        children: [
          {
            id: "image.png",
            icon: "image",
          },
          {
            id: "music.mp3",
            icon: "music_note",
          },
        ],
      },
      {
        id: "lib",
        children: [
          {
            id: "video.mp4",
            icon: "videocam",
          },
          {
            id: "code.tsx",
            icon: "code",
          },
        ],
      },
      {
        id: "child3",
        children: [
          {
            id: "child3-1",
          },
          {
            id: "child3-2",
          },
        ],
      },
    ],
  },
];

function RenderItem({
  item,
  open,
  onOpenChange,
  setItems,
  items,
  map,
  depth,
  onOpenContext,
  selected,
  onClick,
}: RenderProps<Item> & {
  items: Item[];
  selected?: boolean;
  setItems: React.Dispatch<React.SetStateAction<ITree[]>>;
  onOpenContext?: (open: boolean) => void;
  map: Map<string, { item: ITree; parentId?: string }>;
  onClick?: (e: React.MouseEvent) => void;
}) {
  return (
    <ContextMenu
      onOpenChange={onOpenContext}
      content={
        <>
          {item.children !== undefined && (
            <>
              <ContextMenuItem
                onClick={() => {
                  if (item.children) {
                    item.children.push({
                      id: randomId(),
                    });
                  } else {
                    item.children = [
                      {
                        id: `${item.id}-1`,
                      },
                    ];
                  }
                  setItems([...items]);
                }}
              >
                New File
              </ContextMenuItem>
              <ContextMenuItem
                onClick={() => {
                  item.children?.push({
                    id: randomId(),
                    children: [],
                  });

                  setItems([...items]);
                }}
              >
                New Folder
              </ContextMenuItem>
              <ContextMenuSeparator />
            </>
          )}

          <ContextMenuItem
            onClick={() => {
              const parentId = map.get(item.id)?.parentId;
              const parent = map.get(parentId ?? "");
              if (!parent) {
                setItems(items.filter((i) => i.id !== item.id));
                return;
              }
              const index = parent.item.children?.indexOf(item);
              if (index === undefined) throw new Error("index not found");
              parent.item.children?.splice(index, 1);
              if (parent.item.children?.length === 0) parent.item.children = [];
              setItems([...items]);
            }}
          >
            Delete
          </ContextMenuItem>
        </>
      }
    >
      <ListItem
        style={{
          display: "flex",
          alignItems: "center",
          width: "100%",
          height: "100%",
          paddingLeft: (depth + 1) * 8,
          boxSizing: "border-box",
          paddingRight: 8,
          gap: 4,
        }}
        selected={selected}
        onClick={onClick}
        as="div"
      >
        {item.children === undefined ? (
          <Icon
            name={item.icon ?? "note"}
            size={16}
            style={{
              marginLeft: 12,
            }}
          />
        ) : open ? (
          <div
            style={{
              display: "flex",
              gap: 4,
              cursor: "pointer",
            }}
            onClick={() => {
              onOpenChange?.(false);
            }}
          >
            <Icon name="expand_more" size={16} />
            <Icon name="folder_open" size={16} />
          </div>
        ) : (
          <div
            style={{
              display: "flex",
              cursor: "pointer",
              gap: 4,
            }}
            onClick={() => {
              onOpenChange?.(true);
            }}
          >
            <Icon name="chevron_right" size={16} />
            <Icon
              name="folder"
              style={{
                cursor: "pointer",
              }}
              size={16}
            />
          </div>
        )}
        {item.id}
      </ListItem>
    </ContextMenu>
  );
}

export const Basic: Story = {
  render: function Render() {
    const [selectedIds, setSelectedIds] = useState<string[]>([]);
    const [items, setItems] = useState(
      JSON.parse(JSON.stringify(defaultItems)) as ITree[],
    );
    const map = new Map<
      string,
      {
        item: ITree;
        parentId?: string;
      }
    >();
    const traverse = (item: ITree, parentId?: string) => {
      map.set(item.id, {
        item,
        parentId: parentId,
      });
      item.children?.forEach((child) => traverse(child, item.id));
    };
    items.forEach((item) => traverse(item));

    function handleOrderChange(
      src: string,
      dist: string,
      type: OrderType | null,
    ) {
      const srcItem = map.get(src);
      const distItem = map.get(dist);
      if (!srcItem || !distItem) throw new Error("item not found");

      const srcParent = map.get(srcItem.parentId ?? "")?.item;
      const distParent = map.get(distItem.parentId ?? "")?.item;

      if (inChildren(srcItem.item, distItem.item.id)) return;
      if (type === OrderType.Before) {
        if (distParent === undefined) {
          if (srcParent) {
            const srcIndex = srcParent.children?.indexOf(srcItem.item);
            if (srcIndex === undefined) throw new Error("index not found");
            srcParent.children?.splice(srcIndex, 1);
            setItems([srcItem.item, ...items]);
          } else {
            const srcIndex = items.indexOf(srcItem.item);
            const newItems = items.filter((_, index) => index !== srcIndex);
            newItems.unshift(srcItem.item);
            setItems(newItems);
          }
        } else {
          const distIndex = distParent.children?.indexOf(distItem.item);
          if (distIndex === undefined) throw new Error("index not found");

          const srcIndex = srcParent?.children?.indexOf(srcItem.item);
          if (srcIndex === undefined) throw new Error("index not found");
          if (srcParent === undefined)
            throw new Error("srcParent is undefined");

          srcParent.children?.splice(srcIndex, 1);
          distParent.children?.splice(distIndex, 0, srcItem.item);
          setItems([...items]);
        }
      } else if (type === OrderType.After) {
        if (distParent === undefined) {
          if (srcParent) {
            const srcIndex = srcParent.children?.indexOf(srcItem.item);
            if (srcIndex === undefined) throw new Error("index not found");
            srcParent.children?.splice(srcIndex, 1);
            setItems([...items, srcItem.item]);
          } else {
            const srcIndex = items.indexOf(srcItem.item);
            const newItems = items.filter((_, index) => index !== srcIndex);
            newItems.push(srcItem.item);
            setItems(newItems);
          }
        } else {
          const distIndex = distParent.children?.indexOf(distItem.item);
          if (distIndex === undefined) throw new Error("index not found");
          const srcIndex = srcParent?.children?.indexOf(srcItem.item);
          if (srcIndex === undefined) throw new Error("index not found");
          if (srcParent === undefined)
            throw new Error("srcParent is undefined");
          srcParent.children?.splice(srcIndex, 1);
          distParent.children?.splice(distIndex + 1, 0, srcItem.item);
          setItems([...items]);
        }
      } else if (type === OrderType.Child) {
        if (srcParent) {
          const srcIndex = srcParent.children?.indexOf(srcItem.item);
          if (srcIndex === undefined) throw new Error("index not found");

          if (!distItem.item.children) {
            return;
          }
          srcParent.children?.splice(srcIndex, 1);
          distItem.item.children?.push(srcItem.item);
          setItems([...items]);
        } else {
          const srcIndex = items.indexOf(srcItem.item);
          if (srcIndex === undefined) throw new Error("index not found");

          if (!distItem.item.children) {
            return;
          }
          const newItems = items.filter((_, index) => index !== srcIndex);
          distItem.item.children?.push(srcItem.item);
          setItems(newItems);
        }
      }
    }

    return (
      <ContextMenu
        content={
          <>
            <ContextMenuItem
              onClick={() => {
                const newItem = {
                  id: randomId(),
                };
                setItems([...items, newItem]);
              }}
            >
              New File
            </ContextMenuItem>
            <ContextMenuItem
              onClick={() => {
                const newItem = {
                  id: randomId(),
                  children: [],
                };
                setItems([...items, newItem]);
              }}
            >
              New Folder
            </ContextMenuItem>
          </>
        }
      >
        <div
          style={{
            height: 256,
            width: 256,
            overflow: "auto",
          }}
        >
          <TreeView<Item>
            items={items}
            shadowRender={(props) => {
              if (selectedIds.length > 1) {
                const item = map.get(selectedIds[0])?.item;
                return (
                  <div>
                    {item?.id} + {selectedIds.length - 1} more
                  </div>
                );
              }
              const item = map.get(props.item.id)?.item;
              return <div>{item?.id}</div>;
            }}
            render={(props) => (
              <RenderItem
                {...props}
                selected={selectedIds.includes(props.item.id)}
                onOpenContext={(open) => {
                  if (open) {
                    setSelectedIds([props.item.id]);
                  }
                }}
                onClick={(e) => {
                  if (e.metaKey) {
                    setSelectedIds([...selectedIds, props.item.id]);
                    return;
                  }
                  setSelectedIds([props.item.id]);
                }}
                items={items}
                setItems={setItems}
                map={map}
              />
            )}
            onOrderChange={(src, dist, type) => {
              if (selectedIds.length === 1) {
                handleOrderChange(src, dist, type);
                return;
              }
              for (const id of selectedIds) {
                handleOrderChange(id, dist, type);
              }
            }}
          />
        </div>
      </ContextMenu>
    );
  },
};
