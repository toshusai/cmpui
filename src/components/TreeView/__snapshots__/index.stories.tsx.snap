// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`View/TreeView Basic smoke-test 1`] = `
<div data-state="closed"
     style="height: 256px; width: 256px; overflow: auto;"
>
  <div class="cmpui_tree-view__root">
    <ol class="cmpui_tree-view__ol">
      <li class="cmpui_tree-view__item">
        <div class="cmpui_list-item__root"
             aria-selected="false"
             data-size="M"
             data-state="closed"
             style="display: flex; align-items: center; width: 100%; height: 100%; padding-left: 8px; box-sizing: border-box; padding-right: 8px; gap: 4px;"
        >
          <div style="display: flex; gap: 4px; cursor: pointer;">
            <span class="material-symbols-outlined"
                  style="font-size: 16px;"
            >
              expand_more
            </span>
            <span class="material-symbols-outlined"
                  style="font-size: 16px;"
            >
              folder_open
            </span>
          </div>
          root
        </div>
      </li>
      <li class="cmpui_tree-view__item-dir">
        <ol class="cmpui_tree-view__item-dir-ol">
          <li class="cmpui_tree-view__item">
            <div class="cmpui_list-item__root"
                 aria-selected="false"
                 data-size="M"
                 data-state="closed"
                 style="display: flex; align-items: center; width: 100%; height: 100%; padding-left: 16px; box-sizing: border-box; padding-right: 8px; gap: 4px;"
            >
              <div style="display: flex; gap: 4px; cursor: pointer;">
                <span class="material-symbols-outlined"
                      style="font-size: 16px;"
                >
                  expand_more
                </span>
                <span class="material-symbols-outlined"
                      style="font-size: 16px;"
                >
                  folder_open
                </span>
              </div>
              src
            </div>
          </li>
          <li class="cmpui_tree-view__item-dir">
            <ol class="cmpui_tree-view__item-dir-ol">
              <li class="cmpui_tree-view__item">
                <div class="cmpui_list-item__root"
                     aria-selected="false"
                     data-size="M"
                     data-state="closed"
                     style="display: flex; align-items: center; width: 100%; height: 100%; padding-left: 24px; box-sizing: border-box; padding-right: 8px; gap: 4px;"
                >
                  <span class="material-symbols-outlined"
                        style="font-size: 16px; margin-left: 12px;"
                  >
                    image
                  </span>
                  image.png
                </div>
              </li>
              <li class="cmpui_tree-view__item">
                <div class="cmpui_list-item__root"
                     aria-selected="false"
                     data-size="M"
                     data-state="closed"
                     style="display: flex; align-items: center; width: 100%; height: 100%; padding-left: 24px; box-sizing: border-box; padding-right: 8px; gap: 4px;"
                >
                  <span class="material-symbols-outlined"
                        style="font-size: 16px; margin-left: 12px;"
                  >
                    music_note
                  </span>
                  music.mp3
                </div>
              </li>
            </ol>
          </li>
          <li class="cmpui_tree-view__item">
            <div class="cmpui_list-item__root"
                 aria-selected="false"
                 data-size="M"
                 data-state="closed"
                 style="display: flex; align-items: center; width: 100%; height: 100%; padding-left: 16px; box-sizing: border-box; padding-right: 8px; gap: 4px;"
            >
              <div style="display: flex; gap: 4px; cursor: pointer;">
                <span class="material-symbols-outlined"
                      style="font-size: 16px;"
                >
                  expand_more
                </span>
                <span class="material-symbols-outlined"
                      style="font-size: 16px;"
                >
                  folder_open
                </span>
              </div>
              lib
            </div>
          </li>
          <li class="cmpui_tree-view__item-dir">
            <ol class="cmpui_tree-view__item-dir-ol">
              <li class="cmpui_tree-view__item">
                <div class="cmpui_list-item__root"
                     aria-selected="false"
                     data-size="M"
                     data-state="closed"
                     style="display: flex; align-items: center; width: 100%; height: 100%; padding-left: 24px; box-sizing: border-box; padding-right: 8px; gap: 4px;"
                >
                  <span class="material-symbols-outlined"
                        style="font-size: 16px; margin-left: 12px;"
                  >
                    videocam
                  </span>
                  video.mp4
                </div>
              </li>
              <li class="cmpui_tree-view__item">
                <div class="cmpui_list-item__root"
                     aria-selected="false"
                     data-size="M"
                     data-state="closed"
                     style="display: flex; align-items: center; width: 100%; height: 100%; padding-left: 24px; box-sizing: border-box; padding-right: 8px; gap: 4px;"
                >
                  <span class="material-symbols-outlined"
                        style="font-size: 16px; margin-left: 12px;"
                  >
                    code
                  </span>
                  code.tsx
                </div>
              </li>
            </ol>
          </li>
          <li class="cmpui_tree-view__item">
            <div class="cmpui_list-item__root"
                 aria-selected="false"
                 data-size="M"
                 data-state="closed"
                 style="display: flex; align-items: center; width: 100%; height: 100%; padding-left: 16px; box-sizing: border-box; padding-right: 8px; gap: 4px;"
            >
              <div style="display: flex; gap: 4px; cursor: pointer;">
                <span class="material-symbols-outlined"
                      style="font-size: 16px;"
                >
                  expand_more
                </span>
                <span class="material-symbols-outlined"
                      style="font-size: 16px;"
                >
                  folder_open
                </span>
              </div>
              child3
            </div>
          </li>
          <li class="cmpui_tree-view__item-dir">
            <ol class="cmpui_tree-view__item-dir-ol">
              <li class="cmpui_tree-view__item">
                <div class="cmpui_list-item__root"
                     aria-selected="false"
                     data-size="M"
                     data-state="closed"
                     style="display: flex; align-items: center; width: 100%; height: 100%; padding-left: 24px; box-sizing: border-box; padding-right: 8px; gap: 4px;"
                >
                  <span class="material-symbols-outlined"
                        style="font-size: 16px; margin-left: 12px;"
                  >
                    note
                  </span>
                  child3-1
                </div>
              </li>
              <li class="cmpui_tree-view__item">
                <div class="cmpui_list-item__root"
                     aria-selected="false"
                     data-size="M"
                     data-state="closed"
                     style="display: flex; align-items: center; width: 100%; height: 100%; padding-left: 24px; box-sizing: border-box; padding-right: 8px; gap: 4px;"
                >
                  <span class="material-symbols-outlined"
                        style="font-size: 16px; margin-left: 12px;"
                  >
                    note
                  </span>
                  child3-2
                </div>
              </li>
            </ol>
          </li>
        </ol>
      </li>
    </ol>
  </div>
</div>
`;
